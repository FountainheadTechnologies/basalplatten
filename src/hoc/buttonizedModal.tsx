import React from 'react';
import Button from 'antd/lib/button';
import { ButtonProps as AntButtonProps } from 'antd/lib/button/button';
import 'antd/lib/button/style';
import Modal from 'antd/lib/modal';
import { ModalProps as AntModalProps } from 'antd/lib/modal/Modal';
import 'antd/lib/modal/style';

export interface ButtonProps extends AntButtonProps {
  label: string;
}

export interface ModalProps extends AntModalProps {
  closeButton: boolean;
}

export interface Props {
  button?: Partial<ButtonProps>;
  modal?: Partial<ModalProps>;
}

export interface InjectedProps {
  setVisible: (visible: boolean) => void;
  setWorking: (working: boolean) => void;
  setFooter: (footer: React.ReactNode) => void;
  setOkHandler: (okHandler: (() => void)) => void;
}

export interface State {
  modal: {
    visible: boolean;
    working: boolean;
    okHandler?: () => void;
    footer?: React.ReactNode;
  };
  mappedProps: Props;
}

export const buttonizedModal =
  <P extends any>(
    ModalContent: React.ComponentType<P>,
    propMapper: ((props: P) => Props) | Props
  ): React.ComponentClass<P> =>
    class ButtonizedModal extends React.Component<P & Props & InjectedProps, State> {
      state: State = {
        modal: {
          visible: false,
          working: false
        },

        mappedProps: typeof propMapper === 'function' ?
          propMapper(this.props as any) :
          propMapper
      };

      render() {
        const buttonProps = this.buttonProps();

        return (
          <Button {...buttonProps}>
            {buttonProps.label}
            {this.renderModal()}
          </Button>
        );
      }

      buttonProps(): ButtonProps {
        const { mappedProps } = this.state;

        return {
          ...mappedProps.button,
          ...this.props.button as ButtonProps,
          onClick: () => this.setVisible(true)
        }
      }

      renderModal() {
        const { modal } = this.state;

        if (!modal.visible) {
          return null;
        }

        const modalProps = this.modalProps();
        const modalContentProps = this.modalContentProps();

        return (
          <Modal {...modalProps}>
            <ModalContent {...modalContentProps} />
          </Modal>
        );
      }

      modalProps(): Props['modal'] {
        const { modal, mappedProps } = this.state;

        const modalProps: ModalProps = {
          ...mappedProps.modal,
          ...this.props.modal as ModalProps,
          ...modal,

          confirmLoading: modal.working,
          closable: !modal.working,
          onCancel: () => this.setVisible(false),
          onOk: () => this.onOk()
        }

        if (!('maskClosable' in modalProps)) {
          modalProps.maskClosable = !this.state.modal.working;
        }

        if (!modalProps.footer && mappedProps.modal && mappedProps.modal.closeButton) {
          modalProps.footer = (
            <Button size="large" onClick={() => this.setVisible(false)}>
              Close
            </Button>
          );
        }

        return modalProps;
      }

      modalContentProps(): Props & InjectedProps {
        const { mappedProps } = this.state;

        return {
          ...this.props as Props,
          ...mappedProps,
          setVisible: this.setVisible,
          setWorking: this.setWorking,
          setFooter: this.setFooter,
          setOkHandler: this.setOkHandler
        }
      }

      setVisible = (visible: boolean) => {
        const { modal } = this.state;

        this.setState({
          modal: {
            ...modal,
            working: visible ? modal.working : false,
            visible
          }
        });
      }

      setWorking = (working: boolean) => {
        const { modal } = this.state;

        this.setState({
          modal: {
            ...modal,
            working
          }
        });
      }

      setFooter = (footer: React.ReactNode) => {
        const { modal } = this.state;

        this.setState({
          modal: {
            ...modal,
            footer
          }
        });
      }

      setOkHandler = (okHandler: () => void) => {
        const { modal } = this.state;

        this.setState({
          modal: {
            ...modal,
            okHandler
          }
        });
      }

      onOk = () => {
        const { modal } = this.state;

        if (modal.okHandler) {
          return modal.okHandler();
        }

        this.setVisible(false);
      }
    }
