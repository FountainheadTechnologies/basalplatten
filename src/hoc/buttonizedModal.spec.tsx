import React from 'react';
import { shallow } from 'enzyme';

import { buttonizedModal } from './buttonizedModal';

const TestModalContent = () => (
  <div>Test Modal Content</div>
);

describe('default behaviour', () => {
  const TestModal = buttonizedModal(TestModalContent, {
    button: {
      label: 'Test'
    }
  });

  const root = shallow(<TestModal />);

  it('renders a button', () => {
    expect(root).toMatchSnapshot();
    expect(root.find('Button').exists()).toBe(true);
    expect(root.find('Modal').exists()).toBe(false);
  });

  it('displays a modal when button is clicked', () => {
    root.find('Button').simulate('click');
    expect(root.find('Modal').prop('visible')).toBe(true);
  });

  it('hides the modal when modal.onOk is called', () => {
    root.find('Modal').props().onOk();
    expect(root.find('Modal').exists()).toBe(false);
  });

  it('hides the modal when modal.onCancel is called', () => {
    root.find('Button').simulate('click');
    root.find('Modal').props().onCancel();
    expect(root.find('Modal').exists()).toBe(false);
  });

  it('calls the handler set through `setOkHandler` when modal.onOk is called', () => {
    const handler = jest.fn();

    root.instance().setOkHandler(handler);
    root.find('Button').simulate('click');
    root.find('Modal').props().onOk();

    expect(handler).toHaveBeenCalled();
  });

  it('updates the footer when `setFooter` is called', () => {
    root.instance().setFooter('Hello');
    expect(root.find('Modal').prop('footer')).toBe('Hello');
  });

  it('sets the working state when `setWorking` is called', () => {
    root.instance().setWorking(true);
    const modalProps = root.find('Modal').props();

    expect(modalProps.confirmLoading).toBe(true);
    expect(modalProps.closable).toBe(false);
    expect(modalProps.maskClosable).toBe(false);
  });
})

describe('when `modal.closeButton` is `true`', () => {
  const TestModal = buttonizedModal(TestModalContent, {
    button: {
      label: 'Test'
    },
    modal: {
      closeButton: true
    }
  });

  const root = shallow(<TestModal />);

  it('only renders a close button in the modal footer', () => {
    root.find('Button').simulate('click');
    expect(root.find('Modal').prop('footer')).toMatchSnapshot();
  });

  it('closes the modal when button is clicked', () => {
    const button = shallow(root.find('Modal').prop('footer'));
    button.simulate('click');

    expect(root.find('Modal').exists()).toBe(false);
  });
});

describe('when `modal.maskClosable` is set', () => {
  const TestModal = buttonizedModal(TestModalContent, {
    modal: {
      maskClosable: true
    }
  });

  const root = shallow(<TestModal />);

  it('does not change when in working state', () => {
    root.find('Button').simulate('click');
    root.instance().setWorking(true);

    expect(root.find('Modal').prop('maskClosable')).toBe(true);
  });
});

describe('when `propMapper` is a function', () => {
  const TestModal = buttonizedModal(TestModalContent, props => ({
    modal: {
      title: `Test ${props.title}`
    }
  }));

  const root = shallow(<TestModal title="123" />);

  it('calls on mount and uses result to build child props', () => {
    root.find('Button').simulate('click');
    expect(root.find('Modal').prop('title')).toBe('Test 123');
  });
});
