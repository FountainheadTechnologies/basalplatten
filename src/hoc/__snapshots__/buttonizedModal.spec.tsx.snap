// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default behaviour renders a button 1`] = `
<Button
  clicked={false}
  ghost={false}
  label="Test"
  loading={false}
  onClick={[Function]}
  prefixCls="ant-btn"
>
  Test
</Button>
`;

exports[`when \`modal.closeButton\` is \`true\` only renders a close button in the modal footer 1`] = `
<Button
  clicked={false}
  ghost={false}
  loading={false}
  onClick={[Function]}
  prefixCls="ant-btn"
  size="large"
>
  Close
</Button>
`;
